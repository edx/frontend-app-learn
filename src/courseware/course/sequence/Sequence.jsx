/* eslint-disable no-use-before-define */
import React, {
  useEffect, useContext, useState,
} from 'react';
import PropTypes from 'prop-types';
import { sendTrackEvent } from '@edx/frontend-platform/analytics';
import { injectIntl, intlShape } from '@edx/frontend-platform/i18n';
import { useSelector } from 'react-redux';

import PageLoading from '../../../PageLoading';
import { UserMessagesContext, ALERT_TYPES } from '../../../user-messages';
import { useModel } from '../../../model-store';

import messages from './messages';
import { SequenceNavigation, UnitNavigation } from './sequence-navigation';
import SequenceContent from './SequenceContent';

function Sequence({
  unitId,
  sequenceId,
  courseId,
  unitNavigationHandler,
  nextSequenceHandler,
  previousSequenceHandler,
  intl,
}) {
  const sequence = useModel('sequences', sequenceId);
  const unit = useModel('units', unitId);
  const sequenceStatus = useSelector(state => state.courseware.sequenceStatus);
  const handleNext = () => {
    const nextIndex = sequence.unitIds.indexOf(unitId) + 1;
    if (nextIndex < sequence.unitIds.length) {
      const newUnitId = sequence.unitIds[nextIndex];
      handleNavigate(newUnitId);
    } else {
      nextSequenceHandler();
    }
  };

  const handlePrevious = () => {
    const previousIndex = sequence.unitIds.indexOf(unitId) - 1;
    if (previousIndex >= 0) {
      const newUnitId = sequence.unitIds[previousIndex];
      handleNavigate(newUnitId);
    } else {
      previousSequenceHandler();
    }
  };

  const handleNavigate = (destinationUnitId) => {
    unitNavigationHandler(destinationUnitId);
  };

  const logEvent = (eventName, widgetPlacement, targetUnitId) => {
    // Note: tabs are tracked with a 1-indexed position
    // as opposed to a 0-index used throughout this MFE
    const currentIndex = sequence.unitIds.length > 0 ? sequence.unitIds.indexOf(unitId) : 0;
    const payload = {
      current_tab: currentIndex + 1,
      id: unitId,
      tab_count: sequence.unitIds.length,
      widget_placement: widgetPlacement,
    };
    if (targetUnitId) {
      const targetIndex = sequence.unitIds.indexOf(targetUnitId);
      payload.target_tab = targetIndex + 1;
    }
    sendTrackEvent(eventName, payload);
  };

  const { add, remove } = useContext(UserMessagesContext);
  useEffect(() => {
    let id = null;
    if (sequenceStatus === 'loaded') {
      if (sequence.bannerText) {
        id = add({
          code: null,
          dismissible: false,
          text: sequence.bannerText,
          type: ALERT_TYPES.INFO,
          topic: 'sequence',
        });
      }
    }
    return () => {
      if (id) {
        remove(id);
      }
    };
  }, [sequenceStatus, sequence]);

  const [unitHasLoaded, setUnitHasLoaded] = useState(false);
  const handleUnitLoaded = () => {
    setUnitHasLoaded(true);
  };
  useEffect(() => {
    if (unit) {
      setUnitHasLoaded(false);
    }
  }, [unit]);

  if (sequenceStatus === 'loading') {
    if (!sequenceId) {
      return (<div> {intl.formatMessage(messages['learn.sequence.no.content'])} </div>);
    }
    return (
      <PageLoading
        srMessage={intl.formatMessage(messages['learn.loading.learning.sequence'])}
      />
    );
  }

  const gated = sequence && sequence.gatedContent !== undefined && sequence.gatedContent.gated;

  if (sequenceStatus === 'loaded') {
    return (
      <div className="sequence-container">
        <div className="sequence">
          <SequenceNavigation
            sequenceId={sequenceId}
            unitId={unitId}
            className="mb-4"
            nextSequenceHandler={() => {
              logEvent('edx.ui.lms.sequence.next_selected', 'top');
              handleNext();
            }}
            onNavigate={(destinationUnitId) => {
              logEvent('edx.ui.lms.sequence.tab_selected', 'top', destinationUnitId);
              handleNavigate(destinationUnitId);
            }}
            previousSequenceHandler={() => {
              logEvent('edx.ui.lms.sequence.previous_selected', 'top');
              handlePrevious();
            }}
          />
          <div className="unit-container flex-grow-1">
            <SequenceContent
              courseId={courseId}
              gated={gated}
              sequenceId={sequenceId}
              unitId={unitId}
              unitLoadedHandler={handleUnitLoaded}
            />
            {unitHasLoaded && (
              <UnitNavigation
                sequenceId={sequenceId}
                unitId={unitId}
                onClickPrevious={() => {
                  logEvent('edx.ui.lms.sequence.previous_selected', 'bottom');
                  handlePrevious();
                }}
                onClickNext={() => {
                  logEvent('edx.ui.lms.sequence.next_selected', 'bottom');
                  handleNext();
                }}
              />
            )}
          </div>
        </div>
      </div>
    );
  }

  // sequence status 'failed' and any other unexpected sequence status.
  return (
    <p className="text-center py-5 mx-auto" style={{ maxWidth: '30em' }}>
      {intl.formatMessage(messages['learn.course.load.failure'])}
    </p>
  );
}

Sequence.propTypes = {
  unitId: PropTypes.string,
  sequenceId: PropTypes.string,
  courseId: PropTypes.string.isRequired,
  unitNavigationHandler: PropTypes.func.isRequired,
  nextSequenceHandler: PropTypes.func.isRequired,
  previousSequenceHandler: PropTypes.func.isRequired,
  intl: intlShape.isRequired,
};

Sequence.defaultProps = {
  sequenceId: null,
  unitId: null,
};

export default injectIntl(Sequence);
