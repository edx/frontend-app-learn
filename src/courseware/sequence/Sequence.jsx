/* eslint-disable no-use-before-define */
import React, {
  useEffect, useContext, Suspense, useState,
} from 'react';
import PropTypes from 'prop-types';
import { sendTrackEvent } from '@edx/frontend-platform/analytics';
import { injectIntl, intlShape } from '@edx/frontend-platform/i18n';

import Unit from './Unit';
import { SequenceNavigation, UnitNavigation } from '../sequence-navigation';
import PageLoading from '../../PageLoading';
import messages from './messages';
import UserMessagesContext from '../../user-messages/UserMessagesContext';
import { unitShape, sequenceShape, statusShape } from '../course/shapes';

const ContentLock = React.lazy(() => import('./content-lock'));

function Sequence({
  unit,
  sequence,
  status,
  courseUsageKey,
  unitNavigationHandler,
  nextSequenceHandler,
  previousSequenceHandler,
  intl,
}) {
  const handleNext = () => {
    const nextIndex = sequence.unitIds.indexOf(unit.id) + 1;
    if (nextIndex < sequence.unitIds.length) {
      const newUnitId = sequence.unitIds[nextIndex];
      handleNavigate(newUnitId);
    } else {
      nextSequenceHandler();
    }
  };

  const handlePrevious = () => {
    const previousIndex = sequence.unitIds.indexOf(unit.id) - 1;
    if (previousIndex >= 0) {
      const newUnitId = sequence.unitIds[previousIndex];
      handleNavigate(newUnitId);
    } else {
      previousSequenceHandler();
    }
  };

  const handleNavigate = (unitId) => {
    unitNavigationHandler(unitId);
  };

  const logEvent = (eventName, widgetPlacement, targetUnitId) => {
    // Note: tabs are tracked with a 1-indexed position
    // as opposed to a 0-index used throughout this MFE
    const currentIndex = sequence.unitIds.indexOf(unit.id);
    const payload = {
      current_tab: currentIndex + 1,
      id: unit.id,
      tab_count: sequence.unitIds.length,
      widget_placement: widgetPlacement,
    };
    if (targetUnitId) {
      const targetIndex = sequence.unitIds.indexOf(targetUnitId);
      payload.target_tab = targetIndex + 1;
    }
    sendTrackEvent(eventName, payload);
  };

  const { add, remove } = useContext(UserMessagesContext);
  useEffect(() => {
    let id = null;
    if (status.sequence === 'loaded') {
      if (sequence.bannerText) {
        id = add({
          code: null,
          dismissible: false,
          text: sequence.bannerText,
          type: 'info',
          topic: 'sequence',
        });
      }
    }
    return () => {
      if (id) {
        remove(id);
      }
    };
  }, [status.sequence, sequence]);

  const [unitHasLoaded, setUnitHasLoaded] = useState(false);
  const handleUnitLoaded = () => {
    setUnitHasLoaded(true);
  };
  useEffect(() => {
    if (unit) {
      setUnitHasLoaded(false);
    }
  }, [unit]);

  if (status.sequence === 'loading') {
    return (
      <PageLoading
        srMessage={intl.formatMessage(messages['learn.loading.learning.sequence'])}
      />
    );
  }

  const gated = sequence.gatedContent !== undefined && sequence.gatedContent.gated;

  if (status.sequence === 'loaded' && unit) {
    return (
      <div className="sequence">
        <SequenceNavigation
          sequenceId={sequence.id}
          unitId={unit.id}
          className="mb-4"
          nextSequenceHandler={() => {
            logEvent('edx.ui.lms.sequence.next_selected', 'top');
            handleNext();
          }}
          onNavigate={(unitId) => {
            logEvent('edx.ui.lms.sequence.tab_selected', 'top', unitId);
            handleNavigate(unitId);
          }}
          previousSequenceHandler={() => {
            logEvent('edx.ui.lms.sequence.previous_selected', 'top');
            handlePrevious();
          }}
        />
        <div className="unit-container flex-grow-1">
          {gated && (
            <Suspense
              fallback={(
                <PageLoading
                  srMessage={intl.formatMessage(messages['learn.loading.content.lock'])}
                />
              )}
            >
              <ContentLock
                courseUsageKey={courseUsageKey}
                sequenceTitle={sequence.title}
                prereqSectionName={sequence.gatedContent.gatedSectionName}
                prereqId={sequence.gatedContent.prereqId}
              />
            </Suspense>
          )}
          {!gated && (
            <Unit
              key={unit.id}
              id={unit.id}
              onLoaded={handleUnitLoaded}
            />
          )}
          {unitHasLoaded && (
            <UnitNavigation
              sequenceId={sequence.id}
              unitId={unit.id}
              onClickPrevious={() => {
                logEvent('edx.ui.lms.sequence.previous_selected', 'bottom');
                handlePrevious();
              }}
              onClickNext={() => {
                logEvent('edx.ui.lms.sequence.next_selected', 'bottom');
                handleNext();
              }}
            />
          )}
        </div>
      </div>
    );
  }

  // sequence status 'failed' and any other unexpected sequence status.
  return (
    <p className="text-center py-5 mx-auto" style={{ maxWidth: '30em' }}>
      {intl.formatMessage(messages['learn.course.load.failure'])}
    </p>
  );
}

Sequence.propTypes = {
  unit: unitShape,
  sequence: sequenceShape,
  status: statusShape.isRequired,
  courseUsageKey: PropTypes.string.isRequired,
  unitNavigationHandler: PropTypes.func.isRequired,
  nextSequenceHandler: PropTypes.func.isRequired,
  previousSequenceHandler: PropTypes.func.isRequired,
  intl: intlShape.isRequired,
};

Sequence.defaultProps = {
  sequence: undefined,
  unit: undefined,
};

export default injectIntl(Sequence);
