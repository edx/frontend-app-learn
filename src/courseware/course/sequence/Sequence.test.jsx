import React from 'react';
import { fireEvent, waitFor } from '@testing-library/dom';
// eslint-disable-next-line import/no-extraneous-dependencies
import { cloneDeep } from 'lodash';
import { sendTrackEvent } from '@edx/frontend-platform/analytics';
import {
  initialState,
  messageEvent, render, screen, testUnits,
} from '../../../setupTest';
import Sequence from './Sequence';

jest.mock('@edx/frontend-platform/analytics');

describe('Sequence', () => {
  const mockData = {
    unitId: '3',
    sequenceId: '1',
    courseId: '1',
    unitNavigationHandler: () => {},
    nextSequenceHandler: () => {},
    previousSequenceHandler: () => {},
    intl: {},
  };

  it('renders correctly without data', () => {
    const { asFragment } = render(
      <Sequence {...mockData} {...{ unitId: undefined, sequenceId: undefined }} />, { initialState: {} },
    );
    expect(screen.getByText('Loading learning sequence...')).toBeInTheDocument();
    expect(screen.queryByRole('button')).not.toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders correctly for gated content', async () => {
    const { asFragment } = render(<Sequence {...mockData} {...{ sequenceId: '3' }} />);
    expect(screen.getByText('Loading locked content messaging...')).toBeInTheDocument();
    // Only `Previous`, `Next` and `Bookmark` buttons.
    expect(screen.getAllByRole('button').length).toEqual(3);

    const beforeLoadingUnit = asFragment();
    expect(beforeLoadingUnit).toMatchSnapshot();

    window.postMessage(messageEvent, '*');
    await waitFor(() => expect(screen.getByText(/You must complete the prerequisite/)).toBeInTheDocument());
    expect(beforeLoadingUnit).toMatchDiffSnapshot(asFragment());
  });

  it('displays error message on sequence load failure', () => {
    const testState = cloneDeep(initialState);
    testState.courseware.sequenceStatus = 'failed';
    const { asFragment } = render(<Sequence {...mockData} />, { initialState: testState });

    expect(screen.getByText('There was an error loading this course.')).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it('handles loading unit', async () => {
    const { asFragment } = render(<Sequence {...mockData} />);
    expect(screen.getByText('Loading learning sequence...')).toBeInTheDocument();
    // Renders navigation buttons plus one button for each unit.
    expect(screen.getAllByRole('button').length).toEqual(3 + testUnits.length);

    const beforeLoadingUnit = asFragment();
    expect(beforeLoadingUnit).toMatchSnapshot();

    window.postMessage(messageEvent, '*');
    await waitFor(() => expect(screen.queryByText('Loading learning sequence...')).not.toBeInTheDocument());
    // At this point there will be 2 `Previous` and 2 `Next` buttons.
    expect(screen.getAllByRole('button', { name: /previous|next/i }).length).toEqual(4);
    expect(beforeLoadingUnit).toMatchDiffSnapshot(asFragment());
  });

  it('navigates to the previous sequence if the unit is the first in the sequence', async () => {
    sendTrackEvent.mockClear();
    const unitId = '1';
    const sequenceId = '2';
    const previousSequenceHandler = jest.fn();
    render(<Sequence {...mockData} {...{ unitId, sequenceId, previousSequenceHandler }} />);

    const sequencePreviousButton = screen.getByRole('button', { name: /previous/i });
    fireEvent.click(sequencePreviousButton);
    expect(previousSequenceHandler).toHaveBeenCalledTimes(1);
    expect(sendTrackEvent).toHaveBeenCalledTimes(1);
    expect(sendTrackEvent).toHaveBeenCalledWith('edx.ui.lms.sequence.previous_selected', {
      current_tab: Number(unitId), id: unitId, tab_count: testUnits.length, widget_placement: 'top',
    });

    window.postMessage(messageEvent, '*');
    await waitFor(() => expect(screen.queryByText('Loading learning sequence...')).not.toBeInTheDocument());
    const unitPreviousButton = screen.getAllByRole('button', { name: /previous/i })
      .filter(button => button !== sequencePreviousButton)[0];
    fireEvent.click(unitPreviousButton);
    expect(previousSequenceHandler).toHaveBeenCalledTimes(2);
    expect(sendTrackEvent).toHaveBeenCalledTimes(2);
    expect(sendTrackEvent).toHaveBeenNthCalledWith(2, 'edx.ui.lms.sequence.previous_selected', {
      current_tab: Number(unitId), id: unitId, tab_count: testUnits.length, widget_placement: 'bottom',
    });
  });

  it('navigates to the next sequence if the unit is the last in the sequence', async () => {
    sendTrackEvent.mockClear();
    const unitId = String(testUnits.length);
    const sequenceId = '1';
    const nextSequenceHandler = jest.fn();
    render(<Sequence {...mockData} {...{ unitId, sequenceId, nextSequenceHandler }} />);

    const sequenceNextButton = screen.getByRole('button', { name: /next/i });
    fireEvent.click(sequenceNextButton);
    expect(nextSequenceHandler).toHaveBeenCalledTimes(1);
    expect(sendTrackEvent).toHaveBeenCalledWith('edx.ui.lms.sequence.next_selected', {
      current_tab: Number(unitId), id: unitId, tab_count: testUnits.length, widget_placement: 'top',
    });

    window.postMessage(messageEvent, '*');
    await waitFor(() => expect(screen.queryByText('Loading learning sequence...')).not.toBeInTheDocument());
    const unitNextButton = screen.getAllByRole('button', { name: /next/i })
      .filter(button => button !== sequenceNextButton)[0];
    fireEvent.click(unitNextButton);
    expect(nextSequenceHandler).toHaveBeenCalledTimes(2);
    expect(sendTrackEvent).toHaveBeenCalledTimes(2);
    expect(sendTrackEvent).toHaveBeenNthCalledWith(2, 'edx.ui.lms.sequence.next_selected', {
      current_tab: Number(unitId), id: unitId, tab_count: testUnits.length, widget_placement: 'bottom',
    });
  });

  it('navigates to the previous/next unit if the unit is not in the corner of the sequence', () => {
    sendTrackEvent.mockClear();
    const unitNavigationHandler = jest.fn();
    const previousSequenceHandler = jest.fn();
    const nextSequenceHandler = jest.fn();
    render(<Sequence {...mockData} {...{ unitNavigationHandler, previousSequenceHandler, nextSequenceHandler }} />);

    fireEvent.click(screen.getByRole('button', { name: /previous/i }));
    expect(previousSequenceHandler).not.toHaveBeenCalled();
    expect(unitNavigationHandler).toHaveBeenCalledWith(String(Number(mockData.unitId) - 1));

    fireEvent.click(screen.getByRole('button', { name: /next/i }));
    expect(nextSequenceHandler).not.toHaveBeenCalled();
    expect(unitNavigationHandler).toHaveBeenNthCalledWith(2, String(Number(mockData.unitId) + 1));

    expect(sendTrackEvent).toHaveBeenCalledTimes(2);
  });

  it('handles the `Previous` buttons for the first unit in the first sequence', async () => {
    sendTrackEvent.mockClear();
    const unitNavigationHandler = jest.fn();
    const previousSequenceHandler = jest.fn();
    const unitId = '1';
    render(<Sequence
      {...mockData}
      {...{
        unitNavigationHandler, previousSequenceHandler, unitId,
      }}
    />);
    window.postMessage(messageEvent, '*');
    await waitFor(() => expect(screen.queryByText('Loading learning sequence...')).not.toBeInTheDocument());

    screen.getAllByRole('button', { name: /previous/i }).forEach(button => fireEvent.click(button));

    expect(previousSequenceHandler).not.toHaveBeenCalled();
    expect(unitNavigationHandler).not.toHaveBeenCalled();
    expect(sendTrackEvent).not.toHaveBeenCalled();
  });

  it('handles the `Next` buttons for the last unit in the last sequence', async () => {
    sendTrackEvent.mockClear();
    const unitNavigationHandler = jest.fn();
    const nextSequenceHandler = jest.fn();
    const unitId = String(testUnits.length);
    const sequenceId = String(Object.keys(initialState.models.sequences).length);
    render(<Sequence
      {...mockData}
      {...{
        unitNavigationHandler, nextSequenceHandler, unitId, sequenceId,
      }}
    />);
    window.postMessage(messageEvent, '*');
    await waitFor(() => expect(screen.queryByText('Loading learning sequence...')).not.toBeInTheDocument());

    screen.getAllByRole('button', { name: /next/i }).forEach(button => fireEvent.click(button));

    expect(nextSequenceHandler).toHaveBeenCalledTimes(1);
    expect(unitNavigationHandler).not.toHaveBeenCalled();
    expect(sendTrackEvent).toHaveBeenCalledWith('edx.ui.lms.sequence.next_selected', {
      current_tab: Number(unitId), id: unitId, tab_count: testUnits.length, widget_placement: 'top',
    });
  });

  it('handles the navigation buttons for empty sequence', async () => {
    sendTrackEvent.mockClear();
    const testState = cloneDeep(initialState);
    testState.models.sequences['1'].unitIds = [];

    const unitNavigationHandler = jest.fn();
    const previousSequenceHandler = jest.fn();
    const nextSequenceHandler = jest.fn();
    render(<Sequence
      {...mockData}
      {...{
        unitNavigationHandler, previousSequenceHandler, nextSequenceHandler,
      }}
    />, { initialState: testState });
    window.postMessage(messageEvent, '*');
    await waitFor(() => expect(screen.queryByText('Loading learning sequence...')).not.toBeInTheDocument());

    screen.getAllByRole('button', { name: /previous/i }).forEach(button => fireEvent.click(button));
    expect(previousSequenceHandler).toHaveBeenCalledTimes(2);
    expect(unitNavigationHandler).not.toHaveBeenCalled();

    screen.getAllByRole('button', { name: /next/i }).forEach(button => fireEvent.click(button));
    expect(nextSequenceHandler).toHaveBeenCalledTimes(2);
    expect(unitNavigationHandler).not.toHaveBeenCalled();

    expect(sendTrackEvent).toHaveBeenNthCalledWith(1, 'edx.ui.lms.sequence.previous_selected', {
      current_tab: 1, id: mockData.unitId, tab_count: 0, widget_placement: 'top',
    });
    expect(sendTrackEvent).toHaveBeenNthCalledWith(2, 'edx.ui.lms.sequence.previous_selected', {
      current_tab: 1, id: mockData.unitId, tab_count: 0, widget_placement: 'bottom',
    });
    expect(sendTrackEvent).toHaveBeenNthCalledWith(3, 'edx.ui.lms.sequence.next_selected', {
      current_tab: 1, id: mockData.unitId, tab_count: 0, widget_placement: 'top',
    });
    expect(sendTrackEvent).toHaveBeenNthCalledWith(4, 'edx.ui.lms.sequence.next_selected', {
      current_tab: 1, id: mockData.unitId, tab_count: 0, widget_placement: 'bottom',
    });
  });

  it('handles unit navigation button', () => {
    sendTrackEvent.mockClear();
    const unitNavigationHandler = jest.fn();
    const targetUnit = '4';
    render(<Sequence {...mockData} {...{ unitNavigationHandler }} />);

    fireEvent.click(screen.getByRole('button', { name: targetUnit }));
    expect(unitNavigationHandler).toHaveBeenCalledWith(targetUnit);
    expect(sendTrackEvent).toHaveBeenCalledWith('edx.ui.lms.sequence.tab_selected', {
      current_tab: Number(mockData.unitId), id: mockData.unitId, target_tab: Number(targetUnit), tab_count: testUnits.length, widget_placement: 'top',
    });
  });
});
