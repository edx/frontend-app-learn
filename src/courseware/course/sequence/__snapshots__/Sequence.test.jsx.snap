// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sequence displays error message on sequence load failure 1`] = `
<DocumentFragment>
  <p
    class="text-center py-5 mx-auto"
    style="max-width: 30em;"
  >
    There was an error loading this course.
  </p>
</DocumentFragment>
`;

exports[`Sequence handles loading unit 1`] = `
<DocumentFragment>
  <div
    class="sequence-container"
  >
    <div
      class="sequence"
    >
      <nav
        class="sequence-navigation mb-4"
      >
        <button
          class="btn previous-btn"
          type="button"
        >
          <img
            alt="fa-chevron-left"
            class="fa-chevron-left"
            data-testid="icon"
          />
          <span>
            Previous
          </span>
        </button>
        <div
          style="flex-basis: 100%; min-width: 0;"
        >
          <div
            class="sequence-navigation-tabs-container"
          >
            <div
              class="sequence-navigation-tabs d-flex flex-grow-1"
              style=""
            >
              <button
                class="btn"
                title="1"
                type="button"
              >
                <img
                  alt="fa-book"
                  class="fa-book"
                  data-testid="icon"
                />
              </button>
              <button
                class="btn"
                title="2"
                type="button"
              >
                <img
                  alt="fa-book"
                  class="fa-book"
                  data-testid="icon"
                />
              </button>
              <button
                class="btn active"
                title="3"
                type="button"
              >
                <img
                  alt="fa-book"
                  class="fa-book"
                  data-testid="icon"
                />
              </button>
              <button
                class="btn"
                title="4"
                type="button"
              >
                <img
                  alt="fa-book"
                  class="fa-book"
                  data-testid="icon"
                />
              </button>
              <button
                class="btn"
                title="5"
                type="button"
              >
                <img
                  alt="fa-book"
                  class="fa-book"
                  data-testid="icon"
                />
              </button>
              <button
                class="btn"
                title="6"
                type="button"
              >
                <img
                  alt="fa-book"
                  class="fa-book"
                  data-testid="icon"
                />
              </button>
              <button
                class="btn"
                title="7"
                type="button"
              >
                <img
                  alt="fa-book"
                  class="fa-book"
                  data-testid="icon"
                />
              </button>
              <button
                class="btn"
                title="8"
                type="button"
              >
                <img
                  alt="fa-book"
                  class="fa-book"
                  data-testid="icon"
                />
              </button>
              <button
                class="btn"
                title="9"
                type="button"
              >
                <img
                  alt="fa-book"
                  class="fa-book"
                  data-testid="icon"
                />
              </button>
              <button
                class="btn"
                title="10"
                type="button"
              >
                <img
                  alt="fa-book"
                  class="fa-book"
                  data-testid="icon"
                />
              </button>
            </div>
          </div>
        </div>
        <button
          class="btn next-btn"
          type="button"
        >
          <span>
            Next
          </span>
          <img
            alt="fa-chevron-right"
            class="fa-chevron-right"
            data-testid="icon"
          />
        </button>
      </nav>
      <div
        class="unit-container flex-grow-1"
      >
        <div
          class="unit"
        >
          <h2
            class="mb-0 h4"
          >
            3
          </h2>
          <button
            aria-disabled="false"
            aria-live="assertive"
            class="btn pgn__stateful-btn pgn__stateful-btn-state-default btn-link px-1 ml-n1 btn-sm"
            type="button"
          >
            <span
              class="d-flex align-items-center justify-content-center"
            >
              <span
                class="pgn__stateful-btn-icon"
              >
                <img
                  alt="fa-bookmark"
                  class="fa-bookmark"
                  data-testid="icon"
                />
              </span>
              <span>
                Bookmark this page
              </span>
            </span>
          </button>
          <div>
            <div
              class="d-flex justify-content-center align-items-center flex-column"
              style="height: 50vh;"
            >
              <div
                class="spinner-border text-primary"
                role="status"
              >
                <span
                  class="sr-only"
                >
                  Loading learning sequence...
                </span>
              </div>
            </div>
          </div>
          <div
            class="unit-iframe-wrapper"
          >
            <iframe
              allowfullscreen=""
              height="0"
              id="unit-iframe"
              referrerpolicy="origin"
              scrolling="no"
              src="http://localhost:18000/xblock/3?show_title=0&show_bookmark_button=0"
              title="3"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Sequence handles loading unit 2`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -190,40 +190,53 @@
                <span>
                  Bookmark this page
                </span>
              </span>
            </button>
-           <div>
-             <div
-               class=\\"d-flex justify-content-center align-items-center flex-column\\"
-               style=\\"height: 50vh;\\"
-             >
-               <div
-                 class=\\"spinner-border text-primary\\"
-                 role=\\"status\\"
-               >
-                 <span
-                   class=\\"sr-only\\"
-                 >
-                   Loading learning sequence...
-                 </span>
-               </div>
-             </div>
-           </div>
            <div
              class=\\"unit-iframe-wrapper\\"
            >
              <iframe
                allowfullscreen=\\"\\"
-               height=\\"0\\"
+               height=\\"300\\"
                id=\\"unit-iframe\\"
                referrerpolicy=\\"origin\\"
                scrolling=\\"no\\"
                src=\\"http://localhost:18000/xblock/3?show_title=0&show_bookmark_button=0\\"
                title=\\"3\\"
              />
            </div>
+         </div>
+         <div
+           class=\\"unit-navigation d-flex\\"
+         >
+           <button
+             class=\\"btn btn-outline-secondary previous-button mr-2\\"
+             type=\\"button\\"
+           >
+             <img
+               alt=\\"fa-chevron-left\\"
+               class=\\"fa-chevron-left\\"
+               data-testid=\\"icon\\"
+             />
+             <span>
+               Previous
+             </span>
+           </button>
+           <button
+             class=\\"btn btn-outline-primary next-button\\"
+             type=\\"button\\"
+           >
+             <span>
+               Next
+             </span>
+             <img
+               alt=\\"fa-chevron-right\\"
+               class=\\"fa-chevron-right\\"
+               data-testid=\\"icon\\"
+             />
+           </button>
          </div>
        </div>
      </div>
    </div>
  </DocumentFragment>"
`;

exports[`Sequence renders correctly for gated content 1`] = `
<DocumentFragment>
  <div
    class="sequence-container"
  >
    <div
      class="sequence"
    >
      <nav
        class="sequence-navigation mb-4"
      >
        <button
          class="btn previous-btn"
          type="button"
        >
          <img
            alt="fa-chevron-left"
            class="fa-chevron-left"
            data-testid="icon"
          />
          <span>
            Previous
          </span>
        </button>
        <button
          class="btn active"
          title="3"
          type="button"
        >
          <img
            alt="fa-book"
            class="fa-book"
            data-testid="icon"
          />
        </button>
        <button
          class="btn next-btn"
          type="button"
        >
          <span>
            Next
          </span>
          <img
            alt="fa-chevron-right"
            class="fa-chevron-right"
            data-testid="icon"
          />
        </button>
      </nav>
      <div
        class="unit-container flex-grow-1"
      >
        <div>
          <div
            class="d-flex justify-content-center align-items-center flex-column"
            style="height: 50vh;"
          >
            <div
              class="spinner-border text-primary"
              role="status"
            >
              <span
                class="sr-only"
              >
                Loading locked content messaging...
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Sequence renders correctly for gated content 2`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -47,26 +47,31 @@
          </button>
        </nav>
        <div
          class=\\"unit-container flex-grow-1\\"
        >
-         <div>
-           <div
-             class=\\"d-flex justify-content-center align-items-center flex-column\\"
-             style=\\"height: 50vh;\\"
+         <h3>
+           <img
+             alt=\\"fa-lock\\"
+             class=\\"fa-lock\\"
+             data-testid=\\"icon\\"
+           />
+            test-sequence-3
+         </h3>
+         <h4>
+           Content Locked
+         </h4>
+         <p>
+           You must complete the prerequisite: 'test-gated-section' to access this content.
+         </p>
+         <p>
+           <button
+             class=\\"btn btn-primary\\"
+             type=\\"button\\"
            >
-             <div
-               class=\\"spinner-border text-primary\\"
-               role=\\"status\\"
-             >
-               <span
-                 class=\\"sr-only\\"
-               >
-                 Loading locked content messaging...
-               </span>
-             </div>
-           </div>
-         </div>
+             Go To Prerequisite Section
+           </button>
+         </p>
        </div>
      </div>
    </div>
  </DocumentFragment>"
`;

exports[`Sequence renders correctly without data 1`] = `
<DocumentFragment>
  <div>
    <div
      class="d-flex justify-content-center align-items-center flex-column"
      style="height: 50vh;"
    >
      <div
        class="spinner-border text-primary"
        role="status"
      >
        <span
          class="sr-only"
        >
          Loading learning sequence...
        </span>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
